README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

A Flutter package to display and visualize credit card scores, or any other type of score, in a visually appealing way.

## Previews

![preview_1](https://github.com/kartikeyaa-k/score-progress-pretty-display-package/assets/67781046/3011093c-dddd-4d03-8462-9ee6014f7f84)

![preview_2](https://github.com/kartikeyaa-k/score-progress-pretty-display-package/assets/67781046/db286f52-463a-4629-a0e4-a69f691695fb)

![preview_3](https://github.com/kartikeyaa-k/score-progress-pretty-display-package/assets/67781046/c4ffc6d4-a072-466f-9152-f1645f039dfe)


## Features

- **Customizable:** You can customize the appearance of the score visualization, including the color, size, and animation.

- **Animated:** The package includes animations to make the score visualization engaging and interactive.

- **Easy to Use:** Integrating this package into your Flutter project is straightforward and requires minimal setup.

## Installation

Add the following dependency to your `pubspec.yaml` file:

```yaml
dependencies:
  score_progress_pretty_display: ^1.0.0 # Use the latest version
```

## Usage


```dart
PrimaryArcAnimationComponent(
          score: 77,
          maxScore: 100,
          arcHeight: 340,
          arcWidth: 340,
          backgroundArcStrokeThickness: 10,
          progressArcStrokeThickness: 10,
          enableStepperEffect: false,
          isRoundEdges: false,
          minScoreTextFontSize: 30,
          maxScoreTextFontSize: 50,
          isRoundOffScoreWhileProgress: true,
          isRoundOffScore: true,
          showOutOfScoreFormat: true,
          isPrgressCurveFilled: false,
          scoreAnimationDuration: Duration(seconds: 2),
          scoreTextAnimationDuration: Duration(milliseconds: 500),
          scoreTextStyle: TextStyle(fontWeight: FontWeight.normal, height: 1),
          arcBackgroundColor: Colors.black12,
          arcProgressGradientColors: [
            Colors.yellowAccent,
            Colors.greenAccent,
            Colors.green,
          ],
        )
```

## Example 

For a complete example, check the example directory included with this package.

## Contribution Guide

1. Feel free to reach out for feedback/discussions on kartikeya.199231@gmail.com

2. Before submit PR, you need to set githook on your machine by paste and run this command from your terminal at root project 
```
git config core.hooksPath .githooks
```

3. When you add new feature/fix issues, please make sure you update the documentation
```
dart doc .
```
4. If possible, I want to keep this package free from external dependencies.

## Roadmap 

1. I am planning to add new animations in the next release.


CHANGELOG.md
## 1.0.0

* This first release contains : Basic arc progress bar, score text with animation and style customizations

## 1.0.1

* This release contains : Previews of the widget on pub dev, other than this there is no change

## 1.0.2

* This release contains : Better updated previews of the widget on pub dev, other than this there is no change

## 1.0.3

* This release contains : Added actual package dependency inside example, other than this there is no change

## 1.0.4

* This release contains : Added pending documentation, other than this there is no change

## 1.0.5

* This release contains : Updated readme.md with usage example, other than this there is no change

lib/src/primary_arc_animation_component.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:score_progress_pretty_display/src/arc_progress_painter_types/basic_arc_progress_painter.dart';

/// [PrimaryArcAnimationComponent] A widget that displays an animated progress arc with a score text.
///
/// This widget is used to visually represent a score or progress with an animated
/// arc and an optional score text.
class PrimaryArcAnimationComponent extends StatefulWidget {
  /// Creates a [PrimaryArcAnimationComponent].
  ///
  /// The [score] is the final value for the progress arc to stop.
  /// The [maxScore] is the maximum possible score or value.
  /// The [arcHeight] is the height of the progress arc.
  /// The [arcWidth] is the width of the progress arc.
  /// The [backgroundArcStrokeThickness] is the thickness of the background arc.
  /// The [progressArcStrokeThickness] is the thickness of the progress arc.
  ///
  /// Additional parameters allow customization of appearance and behavior:
  /// - [minScoreTextFontSize] : The minimum font size for the score text.
  /// - [maxScoreTextFontSize] : The maximum font size for the score text.
  /// - [arcBackgroundColor] : The background color of the progress arc.
  /// - [arcProgressGradientColors] : List of colors for a gradient effect on the progress arc.
  /// - [enableStepperEffect] : Whether to enable a stepwise animation effect when the score changes.
  /// - [isRoundEdges] : Whether to round the edges of the progress arc.
  /// - [isPrgressCurveFilled] : Whether to fill the progress arc with color.
  /// - [showOutOfScoreFormat] : Whether to show the "/maxScore" format after the score when animation completes.
  /// - [isRoundOffScore] : Whether to round off the score value.
  /// - [isRoundOffScoreWhileProgress] : Whether to round off the score value while the progress is ongoing.
  /// - [scoreAnimationDuration] : The duration of the score animation.
  /// - [scoreTextAnimationDuration] : The duration of the score text animation.
  /// - [scoreTextStyle] : The text style for the score text.
  const PrimaryArcAnimationComponent({
    Key? key,
    required this.score,
    required this.maxScore,
    required this.arcHeight,
    required this.arcWidth,
    required this.backgroundArcStrokeThickness,
    required this.progressArcStrokeThickness,
    this.minScoreTextFontSize = 20,
    this.maxScoreTextFontSize = 35,
    required this.arcBackgroundColor,
    required this.arcProgressGradientColors,
    this.enableStepperEffect = false,
    this.isRoundEdges = false,
    this.isPrgressCurveFilled = false,
    this.showOutOfScoreFormat = true,
    this.isRoundOffScore = true,
    this.isRoundOffScoreWhileProgress = true,
    this.scoreAnimationDuration = const Duration(seconds: 2),
    this.scoreTextAnimationDuration = const Duration(milliseconds: 800),
    this.scoreTextStyle = const TextStyle(
      fontWeight: FontWeight.bold,
      height: 1,
    ),
  }) : super(key: key);

  /// The final value for the progress arc to stop.
  final double score;

  /// The maximum possible score or value.
  final double maxScore;

  /// The height of the progress arc.
  final double arcHeight;

  /// The width of the progress arc.
  final double arcWidth;

  /// The thickness of the background arc.
  final double backgroundArcStrokeThickness;

  /// The thickness of the progress arc.
  final double progressArcStrokeThickness;

  /// The minimum font size for the score text.
  final double minScoreTextFontSize;

  /// The maximum font size for the score text.
  final double maxScoreTextFontSize;

  /// The background color of the progress arc.
  final Color arcBackgroundColor;

  /// List of colors for a gradient effect on the progress arc.
  final List<Color> arcProgressGradientColors;

  /// Whether to enable a stepwise animation effect when the score changes.
  final bool enableStepperEffect;

  /// Whether to round the edges of the progress arc.
  final bool isRoundEdges;

  /// Whether to fill the progress arc with color.
  final bool isPrgressCurveFilled;

  /// Whether to show the "/maxScore" format after the score when animation completes.
  final bool showOutOfScoreFormat;

  /// Whether to round off the score value.
  final bool isRoundOffScore;

  /// Whether to round off the score value while the progress is ongoing.
  final bool isRoundOffScoreWhileProgress;

  /// The duration of the score animation.
  final Duration scoreAnimationDuration;

  /// The duration of the score text animation.
  final Duration scoreTextAnimationDuration;

  /// The text style for the score text.
  final TextStyle scoreTextStyle;

  @override
  State<PrimaryArcAnimationComponent> createState() =>
      _PrimaryArcAnimationComponentState();
}

class _PrimaryArcAnimationComponentState
    extends State<PrimaryArcAnimationComponent> with TickerProviderStateMixin {
  late AnimationController scoreProgressController;
  late Animation<double> scoreProgressAnimation;
  late Tween<double> scoreProgressTween;

  late AnimationController fontSizeController;
  late Animation<double> fontSizeAnimation;
  late Tween<double> fontTween;

  @override
  void initState() {
    assert(
      widget.score <= widget.maxScore,
      'score must be lesser or equal to maxScore',
    );
    assert(
      widget.arcHeight > 0,
      'arc height must be greater than zero',
    );
    assert(
      widget.arcWidth > 0,
      'arc width must be greater than zero',
    );

    scoreProgressController = AnimationController(
      vsync: this,
      duration: widget.scoreAnimationDuration,
    );
    scoreProgressTween = Tween<double>(
      begin: 0,
      end: widget.score,
    );
    scoreProgressAnimation =
        scoreProgressTween.animate(scoreProgressController);
    scoreProgressController.forward();

    scoreProgressController.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        fontSizeController.forward();
      }
    });

    fontSizeController = AnimationController(
      vsync: this,
      duration: widget.scoreTextAnimationDuration,
    );

    fontTween = Tween<double>(
      begin: widget.minScoreTextFontSize,
      end: widget.maxScoreTextFontSize,
    );

    fontSizeAnimation = fontTween.animate(
      CurvedAnimation(
        curve: Curves.easeIn,
        parent: fontSizeController,
      ),
    );

    super.initState();
  }

  @override
  void dispose() {
    scoreProgressController.dispose();
    fontSizeController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: widget.arcHeight / 2 + widget.backgroundArcStrokeThickness,
      width: widget.arcWidth + widget.backgroundArcStrokeThickness,
      // decoration: BoxDecoration(border: Border.all()),
      child: AnimatedBuilder(
        animation: scoreProgressAnimation,
        builder: (context, child) {
          return Stack(
            alignment: Alignment.bottomCenter,
            children: [
              CustomPaint(
                painter: BasicArcProgressPainter(
                  currentValue: scoreProgressAnimation.value,
                  destinatonValue: widget.score,
                  totalValue: widget.maxScore,
                  arcHeight: widget.arcHeight,
                  arcWidth: widget.arcWidth,
                  backgroundArcStrokeThickness:
                      widget.backgroundArcStrokeThickness,
                  progressArcStrokeThickness: widget.progressArcStrokeThickness,
                  arcBackgroundColor: widget.arcBackgroundColor,
                  arcProgressGradientColors: widget.arcProgressGradientColors,
                  enableStepperEffect: widget.enableStepperEffect,
                  isRoundEdges: widget.isRoundEdges,
                  isPrgressCurveFilled: widget.isPrgressCurveFilled,
                ),
              ),
              Align(
                alignment: Alignment.bottomCenter, // Adjust alignment as needed
                child: AnimatedBuilder(
                  animation: fontSizeController,
                  builder: (context, child) {
                    var score = scoreProgressAnimation.value
                        .toStringAsFixed(widget.isRoundOffScore ? 0 : 2);
                    if (scoreProgressController.status !=
                        AnimationStatus.completed) {
                      score = scoreProgressAnimation.value.toStringAsFixed(
                        widget.isRoundOffScoreWhileProgress ? 0 : 2,
                      );
                    }
                    final outOf = (scoreProgressController.status ==
                                AnimationStatus.completed &&
                            widget.showOutOfScoreFormat)
                        ? '/${widget.maxScore.toStringAsFixed(0)}'
                        : '';
                    return Transform(
                      transform: Matrix4.identity()
                        ..rotateY(fontSizeAnimation.value * pi)
                        ..rotateY(fontSizeAnimation.value * pi),
                      alignment: FractionalOffset.center,
                      child: Text(
                        score + outOf,
                        style: widget.scoreTextStyle.copyWith(
                          fontSize: fontSizeAnimation.value,
                        ),
                      ),
                    );
                  },
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}


lib/src/arc_progress_painter_types/basic_arc_progress_painter.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'dart:math' as math;

/// A custom painter for rendering an animated progress arc with customizable properties.
class BasicArcProgressPainter extends CustomPainter {
  /// Creates a [BasicArcProgressPainter] with the specified properties.
  ///
  /// The [currentValue] represents the current progress value.
  /// The [destinatonValue] is the target value for the progress arc.
  /// The [totalValue] represents the maximum possible progress.
  /// The [arcHeight] is the height of the progress arc.
  /// The [arcWidth] is the width of the progress arc.
  /// The [backgroundArcStrokeThickness] is the thickness of the background arc.
  /// The [progressArcStrokeThickness] is the thickness of the progress arc.
  /// The [arcBackgroundColor] is the background color of the progress arc.
  /// The [arcProgressGradientColors] is a list of colors for a gradient effect.
  /// The [enableStepperEffect] enables a stepwise animation effect.
  /// The [isPrgressCurveFilled] determines whether to fill the progress arc.
  /// The [isRoundEdges] specifies whether to round the edges of the arc.
  BasicArcProgressPainter({
    required this.currentValue,
    required this.destinatonValue,
    required this.totalValue,
    required this.arcHeight,
    required this.arcWidth,
    required this.backgroundArcStrokeThickness,
    required this.progressArcStrokeThickness,
    required this.arcBackgroundColor,
    required this.arcProgressGradientColors,
    this.enableStepperEffect = false,
    this.isPrgressCurveFilled = false,
    this.isRoundEdges = true,
  });

  /// The current value of the progress arc.
  final double currentValue;

  /// The destination value for the progress arc to reach.
  final double destinatonValue;

  /// The total value representing the maximum possible progress.
  final double totalValue;

  /// The height of the progress arc.
  final double arcHeight;

  /// The width of the progress arc.
  final double arcWidth;

  /// The thickness of the background arc.
  final double backgroundArcStrokeThickness;

  /// The thickness of the progress arc.
  final double progressArcStrokeThickness;

  /// The background color of the progress arc.
  final Color arcBackgroundColor;

  /// List of colors for a gradient effect on the progress arc.
  final List<Color> arcProgressGradientColors;

  /// Whether to enable a stepwise animation effect when the score changes.
  final bool enableStepperEffect;

  /// Whether to fill the progress arc with color.
  final bool isPrgressCurveFilled;

  /// Whether to round the edges of the progress arc.
  final bool isRoundEdges;

  @override
  void paint(Canvas canvas, Size size) {
    final center =
        Offset(size.width / 2, size.height - backgroundArcStrokeThickness);

    canvas.drawArc(
      Rect.fromCenter(
        center: center / 2,
        width: arcWidth,
        height: arcHeight,
      ),
      pi,
      pi,
      false,
      Paint()
        ..style = PaintingStyle.stroke
        ..color = arcBackgroundColor
        ..strokeCap = isRoundEdges ? StrokeCap.round : StrokeCap.square
        ..strokeWidth = backgroundArcStrokeThickness,
    );
    canvas.saveLayer(
      Rect.fromCenter(
        center: center,
        width: arcWidth + backgroundArcStrokeThickness,
        height: arcHeight + backgroundArcStrokeThickness,
      ),
      Paint(),
    );

    final Gradient gradient = SweepGradient(
      startAngle: 1.25 * math.pi / 2,
      endAngle: 5.5 * math.pi / (enableStepperEffect ? 8 : 2),
      tileMode: TileMode.repeated,
      colors: arcProgressGradientColors,
    );

    final newCenter =
        Offset(size.width / 2, size.height - backgroundArcStrokeThickness / 2);
    canvas.drawArc(
      Rect.fromCenter(center: newCenter, width: arcWidth, height: arcHeight),
      pi,
      // pi * currentValue * (destinatonValue / totalValue) / destinatonValue,
      (pi * currentValue) / totalValue,
      false,
      Paint()
        ..style =
            isPrgressCurveFilled ? PaintingStyle.fill : PaintingStyle.stroke
        ..strokeCap = isRoundEdges ? StrokeCap.round : StrokeCap.square
        ..shader = gradient
            .createShader(Rect.fromLTWH(0.0, 0.0, size.width, size.height))
        ..strokeWidth = progressArcStrokeThickness,
    );
    canvas.restore();
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}


example/lib/features/dashboard/dashboard_screen.dart
import 'package:flutter/material.dart';
import 'package:score_progress_pretty_display/score_progress_pretty_display.dart';

class DashboardScreen extends StatelessWidget {
  const DashboardScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text(
          'Credit Score',
          style: TextStyle(
            fontWeight: FontWeight.bold,
            height: 1,
          ),
        ),
      ),
      body: const Center(
        child: PrimaryArcAnimationComponent(
          score: 77,
          maxScore: 100,
          arcHeight: 340,
          arcWidth: 340,
          backgroundArcStrokeThickness: 10,
          progressArcStrokeThickness: 10,
          enableStepperEffect: false,
          isRoundEdges: false,
          minScoreTextFontSize: 30,
          maxScoreTextFontSize: 50,
          isRoundOffScoreWhileProgress: true,
          isRoundOffScore: true,
          showOutOfScoreFormat: true,
          isPrgressCurveFilled: false,
          scoreAnimationDuration: Duration(seconds: 2),
          scoreTextAnimationDuration: Duration(milliseconds: 500),
          scoreTextStyle: TextStyle(fontWeight: FontWeight.normal, height: 1),
          arcBackgroundColor: Colors.black12,
          arcProgressGradientColors: [
            Colors.yellowAccent,
            Colors.greenAccent,
            Colors.green,
          ],
        ),
      ),
    );
  }
}
